### DISCUSSION TITLE
Key association in implicit many-to-many self-relations

### DISCUSSION THREAD

**Discussion by musjj** [prisma/prisma#24709] [Q&A]
### Question

The following schema:

```prisma
model User {
  id         Int     @id @default(autoincrement())
  name       String?
  followedBy User[]  @relation("UserFollows")
  following  User[]  @relation("UserFollows")
}
```

Will generate the following migration:

```sql
CREATE TABLE "User" (
    id integer DEFAULT nextval('"User_id_seq"'::regclass) PRIMARY KEY,
    name text
);
CREATE TABLE "_UserFollows" (
    "A" integer NOT NULL REFERENCES "User"(id) ON DELETE CASCADE ON UPDATE CASCADE,
    "B" integer NOT NULL REFERENCES "User"(id) ON DELETE CASCADE ON UPDATE CASCADE
);
```

How is `A` and `B` associated with `followedBy` and `following` by the client? Is it based on the declaration order or something else? The docs makes no mention of this.

**Comment by ludralph** [comment]
Hi @musjj 👋 

Thank you for raising this question.

The schema you provided defines a many-to-many self-relation for the `User` model. In the generated migration, the table `_UserFollows` is created to manage this relationship. The columns `A` and `B` in the `_UserFollows` table represent the two sides of the many-to-many relationship.

In Prisma, the association between `A` and `B` with `followedBy` and `following` is not based on the declaration order in the schema. Instead, it is managed by Prisma's internal mechanisms to ensure the correct relationships are maintained. The Prisma Client abstracts away these details, so you don't need to worry about the specific mapping of `A` and `B` to `followedBy` and `following`.

When you use the Prisma Client to query or manipulate the `User` model, you can simply use the `followedBy` and `following` fields as defined in your schema. Prisma will handle the underlying SQL operations to ensure the correct associations are made in the `_UserFollows` table.

For example, to create a user with followers and following users, you can use the following Prisma Client code:
```js
const res = await prisma.user.create({
  data: {
    name: "Raphael",
    followedBy: {
      create: [
        { name: "Jon" },
        { name: "Marc" }
      ]
    },
    following: {
      create: [
        { name: "Niko" },
        { name: "Nurul" }
      ]
    }
  }
});
```
To query a user and include their followers and following users, you can use:
```js
const res = await prisma.user.findFirst({
  include: {
    followedBy: true,
    following: true
  }
});
```

This will give you the correct associations without worrying about the specific columns in the `_UserFollows` table.

**Reply by musjj** [reply]
In my case, I needed to drop down to raw queries to perform certain operations that Prisma doesn't support, so the implementation details here are important to me. Is there any documentation about this or at least a link to the code implementing it?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
